### PR TITLE
PP-1105: qsub: standard input notice

### PR DESCRIPTION
**Pull request by vchlum**
#### Issue-ID

* **[PP-1105](https://pbspro.atlassian.net/browse/PP-1105)**

#### Problem description
* New message printed by qsub in order to inform the user that qsub is waiting for script input.

#### Solution description
* Submitting a job, if the script is empty and the job is not an interactive job and we are in the terminal then print the message.

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [x] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by subhasisb** [comment]
@vchlum i feel that writing to stderr would be better. Please see my comments on the forum on the design.

**Comment by mike0042** [comment]
Forum post is here... http://community.pbspro.org/t/pp-1105-qsub-standard-input-notice/770

In this case I find the message more informational as opposed to a warning. I support using stdout over stderr, but I'm willing to concede the point for the sake of getting the code checked in.

**Comment by subhasisb** [comment]
@mike0042 - Thanks! Sure i can agree that the message is more informational. However, in the interest of not breaking anything that depended on parsing qsub output (which used to be basically a job id only in case of successsful submission, it is better to send the message to stderr. And usually anybody consuming stderr would do so after they know that the problem has errored out, basically by interpreting the exit-status first.

**Comment by prakashcv13** [comment]
@subhasisb, @mike0042, @vchlum,

I am still not satisfied with this going to stderr instead of stdout.
1) qsub will look for the job id only after the script is submitted (with CTRL-Z or CTRL-D) and this message is displayed before the script is provided.
2) This message is not displayed in other scenarios.
3) We should try to do what is more correct, which in this case is that the message goes to stdout as it is informational and not an error/warning vis-a-vis doing something just to prevent a handful of scripts from breaking (which we are not sure if they exist in the first place).

**Comment by mike0042** [comment]
One detail to consider... stderr is generally not buffered while stdout is. If you're trying to send a message to the user and don't want buffering to get in the way, then stderr is the way to go. I was looking for a good technical argument to choose one over the other, and I think this qualifies. My vote is for stderr now.

**Comment by prakashcv13** [comment]
AFAIK, stdout is line buffered (when connected to a terminal) and we do have a "\n" in our printf() call, so buffering will not get in the way in an interactive session.
Moreover, this argument for buffering will be true for most, if not all the printf() calls we have in qsub. Should we be converting them all to print to stderr then?

**Comment by bhroam** [comment]
How about we print to stdout, but only if there is a TTY associated with the qsub?

**Comment by prakashcv13** [comment]
@bhroam, I would say we use setbuf or setvbuf to make stdout unbuffered.

**Comment by vchlum** [comment]
@bhroam There is already a test whether stdin is read from the terminal or not in the commit.

**Comment by prakashcv13** [comment]
@vchlum, right, so there is no way that buffering will come in our way for this printf() call. I was thinking of all the other printf() calls. However, that does not fall within the scope of this ticket.

**Comment by subhasisb** [comment]
Another internet discussion about this: https://stackoverflow.com/questions/4919093/should-i-log-messages-to-stderr-or-stdout

Consider this answer:
_Regular output (the actual result of running the program) should go on stdout, things like you mentioned (e.g. diagnostic, notice, warning, error) on stderr.

If there is no "regular output", I would say that it doesn't really matter which one you choose. You could argue that the logging is the only output, so that should go to stdout. Or you could argue that it is still "exceptional information" which should go to stderr._

Anyways, I am fine with either way.

File: src/cmds/qsub.c [hunk 1]
```diff
@@ -4830,6 +4830,15 @@ main(int argc, char **argv, char **envp)   /* qsub */
 	if (command_flag == 0) {
 		/* if script is empty, get standard input */
 		if ((strcmp(script, "") == 0) || (strcmp(script, "-") == 0)) {
+			/* if this is a terminal, print a short info */
+			if (isatty(STDIN_FILENO) && Interact_opt == FALSE) {
```
Comment (bhroam) [inline src/cmds/qsub.c]: I did a little bit of internet sleuthing and it looks like isattr() is deprecated on windows(https://msdn.microsoft.com/en-us/library/ms235388.aspx).  I don't know if this means we should use it or not.  It's still a perfectly valid function in the UNIX world (http://pubs.opengroup.org/onlinepubs/9699919799/).  Should we use isatty() anyway or put it in an ifdef block and use _isatty() on windows and isatty() on linux?

Comment (hirenvadalia) [reply src/cmds/qsub.c]: @bhroam We are using _isatty() in windows... see this https://github.com/PBSPro/pbspro/blob/master/src/include/win.h#L70

Comment (bhroam) [reply src/cmds/qsub.c]: Thanks for pointing that out @hirenvadalia.  I do not see us including win.h in qsub.  Maybe its needed?

Comment (hirenvadalia) [reply src/cmds/qsub.c]: @bhroam It is getting included via windows version of pbs_config.h (see this https://github.com/PBSPro/pbspro/blob/master/win_configure/include/pbs_config.h#L83)

**Comment by vchlum** [comment]
@PBSPro/pbspro-maintainers I have two sign-offs and I rebased the branch. This commit is ready for merge.

**Pull request merged**